lsp-completion.el

```diff
@@ -0,0 +1,553 @@
+
+        (setq-local lsp-inhibit-lsp-hooks nil)
+
+        (when (lsp-completion--looking-back-trigger-characterp trigger-chars)
+          (setq this-command 'self-insert-command)))
+    (lsp-completion--capf-clear-cache)))
+
+(defun lsp-completion--setup-company ()
+  "Setup company-mode."
+  (cond
+   ((and (functionp 'company-lsp)
+         (not lsp-completion-prefer-capf))
+    (progn
+      (company-mode 1)
+      (add-to-list 'company-backends 'company-lsp)
+      (setq-local company-backends (remove 'company-capf company-backends))))
+
+   ((and (fboundp 'company-mode) lsp-completion-enable)
+    (company-mode 1)
+    (add-to-list 'company-backends 'company-capf)))
+
+  (add-hook 'company-completion-started-hook
+            (lambda (&rest _)
+              (setq-local lsp-inhibit-lsp-hooks t))
+            nil
+            t)
+  (add-hook 'company-after-completion-hook
+            (lambda (&rest _)
+              (lsp-completion--capf-clear-cache)
+              (setq-local lsp-inhibit-lsp-hooks nil))
+            nil
+            t))
+
+(defun lsp-completion--clean-company ()
+  "Clean company-mode."
+  (remove-hook 'company-completion-started-hook
+               (lambda (&rest _)
+                 (setq-local lsp-inhibit-lsp-hooks t))
+               t)
+  (remove-hook 'company-after-completion-hook
+               (lambda (&rest _)
+                 (lsp-completion--capf-clear-cache)
+                 (setq-local lsp-inhibit-lsp-hooks nil))
+               t))
+
+(defun lsp-completion--enable ()
+  "Enable LSP completion support."
+  (when (and lsp-completion-enable
+             (lsp-feature? "textDocument/completion"))
+    (setq-local completion-at-point-functions nil)
+    (add-hook 'completion-at-point-functions #'lsp-completion-at-point nil t)
+    (setq-local completion-category-defaults
+                (add-to-list 'completion-category-defaults '(lsp-capf (styles basic))))
+    (lsp-completion-mode 1)))
+
+(defun lsp-completion--disable ()
+  "Disable LSP completion support."
+  (lsp-completion-mode -1)
+  (remove-hook 'completion-at-point-functions #'lsp-completion-at-point t)
+  (setq-local completion-category-defaults
+              (cl-remove 'lsp-capf completion-category-defaults :key #'car)))
+
+;;;###autoload
+(define-minor-mode lsp-completion-mode
+  "Toggle LSP completion support."
+  :group 'lsp-mode
+  :global nil
+  :lighter ""
+  (cond
+   (lsp-completion-mode
+    (add-hook 'lsp-configure-hook #'lsp-completion--enable nil t)
+    (add-hook 'lsp-unconfigure-hook #'lsp-completion--disable nil t)
+    (when (bound-and-true-p company-mode)
+      (lsp-completion--setup-company)))
+   (t
+    (remove-hook 'lsp-configure-hook #'lsp-completion--enable t)
+    (remove-hook 'lsp-unconfigure-hook #'lsp-completion--disable t)
+    (when (bound-and-true-p company-mode)
+      (lsp-completion--clean-company)))))
+
+;;;###autoload
+(add-hook 'lsp--auto-configure-hook #'lsp-completion--enable)
```

Comment:
why do we have to go like this?

Reply:
This way we can handle the auto configure function to it feature and its own file

Reply:
Do you see any problem with this way?

Reply:
lsp-completion--enable is not autoloaded so this most likely will cause trouble. I don't like having executable code on top level in a file. Why do we need two hooks `lsp--auto-configure-hook` and `lsp-configure-hook`? What is different here compared to the rest of the stuff that we moved to separate files?

Reply:
Yeah, I changed to use another autoloaded function which will be called only on autoconfigure.
AFAIK auto-configure is different from configure, they are triggered in different places, one on `lsp` function and another on `lsp-managed-mode`.I think keep this hook of auto-configure for each feature could help organize better the features for it file

Reply:
> AFAIK auto-configure is different from configure,

I think that this is incidental - the code that was in lsp--auto-configure in the old source could be easily moved to managed-mode hook.

Reply:
> the code that was in lsp--auto-configure

By "code" I mean - "the completion related code"

Reply:
So, can I move that code to be enabled when `lsp-configure-hook` run?

Reply:
I think yes. Moreover, this code is incorrect - it should check if there is completionProvider and enable capf only if it is present. And the correct place to check if xxxProvider is present is in managed mode hooks.

Reply:
Done, could you check it again?

Reply:
lsp--auto-configure-hook is no longer needed?

Reply:
Well, without it, we still need to couple `lsp-xyz-enable` with `lsp-mode.el` and add:
```elisp
(when lsp-completion-enable
    (add-hook 'lsp-configure-hook 'lsp-completion-mode))
``` 
in `lsp--auto-configure` function

Reply:
The idea behind lsp-xyz-enable was to prevent loading lsp-xyz.el . If we change all submodules to attach to that private hook we will eagerly load all features even if they are disabled.

Reply:
well, if that is the problem we can wrap checking if `lsp-xyz-enable` is `t` before enabling the submodule, what do you think?
Something like:
```elisp

(defun lsp-completion--enable ()
  "Enable LSP completion support."
    (lsp-completion-mode 1))

;;;###autoload
(defun lsp-completion--auto-enable ()
  (when lsp-completion-enable
    (lsp-completion--enable)))

;;;###autoload
(add-hook 'lsp--auto-configure-hook #'lsp-completion--auto-enable)
```

Reply:
> I think that this is incidental - the code that was in lsp--auto-configure in the old source could be easily moved to managed-mode hook.

Is this correct?
I saw code that change user configuration in `lsp-auto-configure` like setting `company-backends`. So auto configure is definitely different from configuring feature.

Reply:
> I saw code that change user configuration in `lsp-auto-configure` like setting `company-backends`. So auto configure is definitely different from configuring feature.

@kiennq 
Changing company-backends even before the server has sent initialized response doesn't seem the right thing.

Reply:
> well, if that is the problem we can wrap checking if `lsp-xyz-enable` is `t` before enabling the submodule, what do you think?

@ericdallo that won't help - once you call an autoloaded function from a module, the whole module is loaded.

Reply:
Hum, I see, well, I think it's a tradeoff, right? 
1. Load the submodule even if the user doesn't use it.
2. coupling submodule with `lsp-mode.el` without loading the submodule if not enabled.

I still prefer the first one as the most lsp features are default `t` and it looks a cleaner code for me.

Reply:
> * Load the submodule even if the user doesn't use it.

Not only loading module: loading module + all submodule dependencies. E. g. if someone does not use flymake it will still have it loaded because we will load lsp-flymake unconditionally. 

> 2\. coupling submodule with `lsp-mode.el` without loading the submodule if not enabled.

If we aim to what lsp-dart.el is - this is not a concern. Also, we have so much stuff related to integration with external components like evil, projectile, which-key, avy, etc to be concerned with this issue.

Reply:
I see your point, I'm just not sure what is better, so should I remove this `lsp--auto-configure-hook` and go back to what we have on master for the features, what do you think @yyoncho @kiennq?

For completion, I think I'll need to move the content of `lsp-completion--auto-configure` to `lsp-mode.el`? :pensive:

Reply:
> Changing company-backends even before the server has sent initialized response doesn't seem the right thing.

@yyoncho 
That means the location of `auto-configure` should be moved after server initialized response, doesn't it?
What I mean is that the functionality of `auto-configure` and `configure` is different. While `auto-configure` can override user configuration (like `company-backends`), `configure` should not do that.

Reply:
> that won't help - once you call an autoloaded function from a module, the whole module is loaded.

That can be done in easier way while doesn't load the whole module
``` elisp
;;;###autoload
(add-hook 'lsp--auto-configure-hook (lambda () (when lsp-completion-enable (lsp-completion--enable))))
```

No autoloaded function from module is called when `lsp-completion-enable` is `nil`

Reply:
> That means the location of `auto-configure` should be moved after server initialized response, doesn't it?

@kiennq 

Yes. This is what I am saying. `auto-configure` will just set the stuff to run in the `configure` method. I think that you did one pass with that.

> If yes, for completion, I think I'll need to move the content of lsp-completion--auto-configure to lsp-mode.el

@ericdallo why?

Reply:
Oh, @kiennq I thought it was the same of:
```elisp
;;;###autoload
(add-hook 'lsp--auto-configure-hook #'lsp-completion--auto-configure)
```

When we use a lambda for that, emacs won't load the module if is not `t`?

Reply:
> @ericdallo why?
@yyoncho
Where should it go the content of ` lsp-completion--auto-configure` so?

Reply:
In `configure` hook.

Reply:
But then I think we'll have this problem, right?
https://github.com/emacs-lsp/lsp-mode/pull/1967#discussion_r460557714

Reply:
> When we use a lambda for that, emacs won't load the module if is not `t`?

No, since it will not call any autoloaded function.

Reply:
> But then I think we'll have this problem, right?
> [#1967 (comment)](https://github.com/emacs-lsp/lsp-mode/pull/1967#discussion_r460557714)

I don't understand that comment. @kiennq  can you elaborate?

> That's user configuration since it modifies company-backends. I have a special configuration for capf (like multi backend for company-capf and dabbrev-code) and by this, I will have no way of keeping my company-backends unchanged. It will also break my configuration.

You can disable lsp-completion-enable or use lsp-managed-mode-hook to set whatever backends you want.

Reply:
> You can disable lsp-completion-enable or use lsp-managed-mode-hook to set whatever backends you want.

@yyoncho Disable `lsp-completion-enable` is no good, since I want to use `capf`.
Using `lsp-managed-hook` to reset the backends is not good either. That means I have to copy my whole big `company-backends` in two places, one is configuration for `company-mode`, and another one is configuration for `lsp-mode`, that's redundant and doesn't isolate `lsp-mode` from `company-mode`.

My point is `company-backends` is user configuration, and I don't think it's good for other package to mess with it.
`lsp-mode` provide `auto-configure` to sweeten user configuration so it can do that while auto configure is enabled, else it shouldn't do that.

Reply:
> @yyoncho Disable lsp-completion-enable is no good, since I want to use capf.

I don't understand - If you use capf, enabling it is one line.

> My point is `company-backends` is user configuration, and I don't think it's good for other package to mess with it.
> `lsp-mode` provide `auto-configure` to sweeten user configuration so it can do that while auto configure is enabled, else it shouldn't do that.

And that is what will happen - if you disable auto-configuration it won't touch company-backends and it won't perform any configuration so you can configure everything by yourself.

Reply:
To elaborate a bit more, our configuration story is:

1. We provide some autoconfiguration OOTB which targets the majority of the users.
2. If the user does not want autoconfiguration - they can turn everything off via lsp-auto-configure 
3. If the user does not like particular setting - they set lsp-FEATURE-enable to nil
4. If the user does not like how a particular feature is configured - they turn it off, copy the parts that they like, remove the parts that they dont like. 


Regarding 4: most of the features are enabled within 2-5 lines. If there is more code, then it will make sense to introduce an additional variable like lsp-FEATURE-SUBFEATURE-enable to control that. But we cannot introduce a flag to turn off/on every single line of code in the autoconfiguration section.

Reply:
I agree with all these points

Reply:
> And that is what will happen - if you disable auto-configuration it won't touch company-backends and it won't perform any configuration so you can configure everything by yourself.

I think this is too extreme.
Let's say I disable auto-configuration, do you mean that I need to configure `completion-at-point-functions`  ... by myself? Which is require knowledge both about `lsp-mode` and `completion` itself, so I don't think that's easy for user to do.
User can indeed copy those configuration lines, but it calls internal functions and there's no warrantee it will not be broken by next update.
That's why I want to separate `auto-config` from normal configuration. For example, setting `completion-at-point-functions`, `completion-category-defaults` are normal configuration, even `major-mode` already do that and should be enabled when `lsp-mode` supports it.
On other hand, `company-backends` configuration are `auto-config`, when user use `company-mode`, they probably already customize that value already.
If you want to ask where to draw the line, I think it's depend, when we want to configure optional packages that doesn't have hard dependency of `lsp-mode`, that probably should go to `auto-config`?
Does that make sense?

> 4\. If the user does not like how a particular feature is configured - they turn it off, copy the parts that they like, remove the parts that they dont like.

While I agree about configuration stories, the previous implementation that using dual check of `lsp-auto-enable` inside `lsp-configure-buffer` prevent user to run `lsp-configure-hook` when `lsp-auto-enable` is not turned on. Which is quite bad since user cannot turn on feature now.
However, if we remove `lsp-auto-enable` in `lsp-configure-hook`, all of the features `lsp-xyz-enable` which is set to `t` by default will be loaded immediately when buffer is configured.
Is there solution for that without introducing new configuration flags?

I proposed https://github.com/emacs-lsp/lsp-mode/pull/1967#discussion_r460555441 which allows user to disable a feature by setting `lsp-xyz-enable` to `nil` and put a property of `not-autoconfig` on it.

Alternatively, how about setting default value for `lsp-xyz-enable` as `init`, so we know that user haven't customize that value yet, and can do something like enable that feature if `lsp-auto-enable` or set that feature to default enablement state when `lsp-auto-enable` is not enabled?